### PR TITLE
[Fix #197]: Add Sample Objects query

### LINKED ISSUE #197: Add Sample Objects query
It's nice sometimes to be able to select a sampling of objects from some object set; for example, give me a random sampling of 10,000 objects. Use cases include when running merge shortest paths to GC roots on some class results in an OOME but running on a random sample would be likely sufficient.

### PR DESCRIPTION
**Pull request by kgibm**
image

image

Fixes #197

### CODE REVIEW

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    @Argument(flag = Argument.UNFLAGGED)
+    public IHeapObjectArgument objects;
+
+    @Argument(isMandatory = true)
+    public SamplingType sample = SamplingType.RANDOM;
+
+    @Argument(isMandatory = true)
+    public int number = 10000;
+
+    public enum SamplingType
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java:51]: I like this approach of allowing many sample types.

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        {
+            finalObjectIds = new int[number];
+            System.arraycopy(objectIds, objectIds.length - number - 1, finalObjectIds, 0, number);
+        }
+        else if (sample == SamplingType.LARGEST_RETAINED)
+        {
+            finalObjectIds = Arrays.stream(objectIds).boxed().sorted((x, y) -> {
+                try
+                {
+                    return Long.valueOf(snapshot.getRetainedHeapSize(y)).compareTo(snapshot.getRetainedHeapSize(x));
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java:112]: To avoid extra boxing, we can use the form:

```
return Long.compare(snapshot.getRetainedHeapSize(y), snapshot.getRetainedHeapSize(x));
```

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    public ISnapshot snapshot;
+
+    @Argument(flag = Argument.UNFLAGGED)
+    public IHeapObjectArgument objects;
+
+    @Argument(isMandatory = true)
+    public SamplingType sample = SamplingType.RANDOM;
+
+    @Argument(isMandatory = true)
+    public int number = 10000;
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java:49]: Not sure if this is validated in UI? I think it needs to be checked as >= 0.

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        }
+        else if (sample == SamplingType.FIRST)
+        {
+            finalObjectIds = new int[number];
+            System.arraycopy(objectIds, 0, finalObjectIds, 0, number);
+        }
+        else if (sample == SamplingType.LAST)
+        {
+            finalObjectIds = new int[number];
+            System.arraycopy(objectIds, objectIds.length - number - 1, finalObjectIds, 0, number);
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java:105]: This looks like an off by one, eg:

[5 6 7 8], length = 4, number = 1, target = 3.

```
System.arraycopy(objectIds, objectIds.length - number - 1, finalObjectIds, 0, number);
System.arraycopy(objectIds, 4 - 1 - 1, finalObjectIds, 0, 1);
System.arraycopy(objectIds, 2, finalObjectIds, 0, 1);
```
So I think this will get us [7], but I expect we want [8]?

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            return label;
+        }
+    }
+
+    public IResult execute(IProgressListener listener) throws Exception
+    {
+        int[] objectIds = objects.getIds(listener);
+
+        int[] finalObjectIds = null;
+        if (objectIds.length < number)
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java:93]: could this be `<=` ? e.g  length = 0, number = 0 ;; or length = 5, number = 5, could all match

File: plugins/org.eclipse.mat.tests/src/org/eclipse/mat/tests/snapshot/GeneralSnapshotTests.java [hunk 6]
```diff
@@ -1449,6 +1449,17 @@ public void selectObjectsByField() throws SnapshotException
         assertThat(result.getElements().size(), greaterThan(0));
     }
 
+    @Test
+    public void sampleObjects() throws SnapshotException
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.tests/src/org/eclipse/mat/tests/snapshot/GeneralSnapshotTests.java:1453]: Maybe we can add extra test for the LAST.

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            return icon;
+        }
+
+        public String toString()
+        {
+            return label;
+        }
+    }
+
+    public IResult execute(IProgressListener listener) throws Exception
```
Comment (jasonk000) [inline plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java:88]: Could add a try/finally around the whole fn to unwrap and rethrow the wrapped SnapshotException? An extra method layer to wrap the query implementation can keep it tidy.

**Review by jasonk000 — CHANGES_REQUESTED** [review]
Small changes here, and one bug. Direction is good.

File: plugins/org.eclipse.mat.tests/src/org/eclipse/mat/tests/snapshot/GeneralSnapshotTests.java (hunk 6 above)
Comment (jasonk000) [reply plugins/org.eclipse.mat.tests/src/org/eclipse/mat/tests/snapshot/GeneralSnapshotTests.java:1453]: And some more .. the '0', '-ve', and other boundary conditions.

File: plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java (hunk 7 above)
Comment (kgibm) [reply plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java]: I initially tried that but I was surprised to find it didn't work because `getRetainedHeapSize` and `getHeapSize` throw the checked exception `SnapshotException` but `sorted` doesn't throw that exception so handling outside of `sorted` doesn't work. I'm still a bit curmudgeony and haven't fully mentally migrated to closures in Java but this seems like a huge oversight in the language that could have made closures much cleaner.

Comment (jasonk000) [reply plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java]: > handling outside of sorted doesn't work

Good point!

> huge oversight in the language that could have made closures much cleaner

Yes, checked exceptions are a headache. Basically we need to create a `@FunctionalInterface` single-method interface that has the SnapshotException in the signature.

(not critical for change but may be some inspiration!, LLM whipped it up)

```
@FunctionalInterface
private interface SnapshotLongFunction<T>
{
    long applyAsLong(T value) throws SnapshotException;
}
```

```
private static <T> Comparator<T> comparingSnapshot(SnapshotLongFunction<T> function,
                boolean reversed)
{
    return (left, right) -> {
        try
        {
            long l = function.applyAsLong(left);
            long r = function.applyAsLong(right);
            return reversed ? Long.compare(r, l) : Long.compare(l, r);
        }
        catch (SnapshotException e)
        {
            throw new RuntimeException(e);
        }
    };
}
```

Now we can call:
```
...
.sorted(comparingSnapshot(snapshot::getRetainedHeapSize, true))
```

Comment (kgibm) [reply plugins/org.eclipse.mat.api/src/org/eclipse/mat/inspections/SampleObjectsQuery.java]: Interesting. Speaking of which, I'm not sure what the policy is for LLMs for Eclipse and/or MAT? I'm still a bit curmudgeonly with LLMs, too, so I haven't even looked into the policies yet
